Extending execution context for retries

alrod opened the pull request:
resolves #373

alrod committed:
Adding retry properties to executionContext

mhoeger commented on azure-functions-language-worker-protobuf/src/proto/FunctionRpc.proto:
```diff
@@ -263,6 +263,12 @@ message InvocationRequest {
   RpcTraceContext trace_context = 5;
+
+  // current retry count
+  int32 retry_count = 6;
```
The default value for retry_count and max_retry_count will be 0 (because it is a [proto3 numeric type](https://developers.google.com/protocol-buffers/docs/proto3#default)). So even if it isn't explicitly set, the customer will see `context.executionContext.maxRetryCount` => 0 and `context.executionContext.retryCount` => 0. 

I think this is ok and desirable behavior but I just want to double check!

If not, please use [NullableDouble](https://github.com/Azure/azure-functions-language-worker-protobuf/blob/821cff191bd59ec224ce2886af5df039fedbd997/src/proto/shared/NullableTypes.proto#L14) or create and use some NullableInt

alrod replied:
I wrapped retry properties into `RetryContext`. Exactly the same we had in C# `ExecutionContext`.

mhoeger commented on src/public/Interfaces.ts:
```diff
@@ -177,6 +177,14 @@ export interface ExecutionContext {
      */
     functionDirectory: string;
+    /**
+     * Current retry count.
```
Could you add a bit about how to configure retry count? Or a link to docs?

mhoeger commented on src/public/Interfaces.ts:
```diff
@@ -177,6 +177,14 @@ export interface ExecutionContext {
+     */
+    retryCount: number;
+    /**
+     * Current retry count.
```
Same for this comment

mhoeger reviewed (commented):
A few comments, mostly looks good!!

alrod committed:
Adding retryContext

mhoeger reviewed (approved):
LGTM! Thanks for the changes and for the interface comments!!

Before merging this PR though, could you please merge this one: https://github.com/Azure/azure-functions-language-worker-protobuf/pull/53#pullrequestreview-586075083
And then pull updates from azure-functions-language-worker-protobuf to this repo following [these steps](https://github.com/Azure/azure-functions-language-worker-protobuf#pulling-updates)? These steps make sure that proto files don't get out of sync between workers.

Will approve again if that change dismisses that review!

alrod committed:
<>Adding RpcException

alrod committed:
Updated subtree from https://github.com/azure/azure-functions-language-worker-protobuf and added Exception interface

alrod commented:
> LGTM! Thanks for the changes and for the interface comments!!
> 
> Before merging this PR though, could you please merge this one: [Azure/azure-functions-language-worker-protobuf#53 (review)](https://github.com/Azure/azure-functions-language-worker-protobuf/pull/53#pullrequestreview-586075083)
> And then pull updates from azure-functions-language-worker-protobuf to this repo following [these steps](https://github.com/Azure/azure-functions-language-worker-protobuf#pulling-updates)? These steps make sure that proto files don't get out of sync between workers.
> 
> Will approve again if that change dismisses that review!

1. I updated the prorofile from protobuf repository.
2. Also I added new Exception interface for retry context. (it is the same to `IRpcException`)
@mhoeger, please review once again.

mhoeger reviewed (approved):
LGTM!✅
